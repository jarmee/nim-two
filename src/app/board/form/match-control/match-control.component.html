<input
  #matchCheckbox
  data-test-id="match-checkbox"
  type="checkbox"
  (change)="onChange($event.target.checked)"
  (blur)="onTouched()"
  [name]="formControlName"
  [attr.id]="formControlName"
/>
