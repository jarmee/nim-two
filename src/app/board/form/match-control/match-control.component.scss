:host {
  display: block;
  cursor: pointer;

  .custom-control-label {
    cursor: pointer;
    text-align: center;
    &:before {
      content: 'ðŸ’£';
      border: 0px solid transparent;
      background: transparent;
      width: 50px;
      height: 50px;
      line-height: 50px;
    }
    &:after {
      width: 50px;
      height: 50px;
      line-height: 50px;
    }
  }
  .custom-control-input:checked ~ {
    .custom-control-label {
      &:before {
        content: 'ðŸ’¥';
        border: 0px solid transparent;
        background: transparent;
        width: 50px;
        height: 50px;
        line-height: 50px;
      }
      &:after {
        width: 50px;
        height: 50px;
        line-height: 50px;
        background: transparent;
      }
    }
  }
}
