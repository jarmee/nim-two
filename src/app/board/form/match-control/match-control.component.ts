import {
  Component,
  ElementRef,
  forwardRef,
  Input,
  Renderer2,
  ViewChild
} from "@angular/core";
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from "@angular/forms";

@Component({
  selector: "app-match-control",
  templateUrl: "./match-control.component.html",
  styleUrls: ["./match-control.component.scss"],
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => MatchControlComponent),
      multi: true
    }
  ]
})
export class MatchControlComponent implements ControlValueAccessor {
  @Input()
  formControlName: string;

  @Input()
  id: string;

  @ViewChild("matchCheckbox", { static: true })
  checkboxElementRef: ElementRef;

  onChangeCallback: (value: boolean) => void;

  onTouchedCallback: () => void;

  get uniqueFormControlId() {
    return `${this.formControlName}-${this.id}`
  }

  constructor(private renderer: Renderer2, private elementRef: ElementRef) {}

  writeValue(value: boolean): void {
    this.renderer.setProperty(
      this.checkboxElementRef.nativeElement,
      "checked",
      value
    );
    this.renderer.setProperty(
      this.checkboxElementRef.nativeElement,
      "disabled",
      value
    );
  }

  registerOnChange(callback: (value: boolean) => void): void {
    this.onChangeCallback = callback;
  }

  registerOnTouched(callback: () => void): void {
    this.onTouchedCallback = callback;
  }

  onChange(value: boolean) {
    this.onChangeCallback(value);
  }

  onTouched() {
    this.onTouchedCallback();
  }
}
