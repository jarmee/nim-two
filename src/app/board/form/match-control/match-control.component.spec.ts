import { ElementRef, Renderer2, Type } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import * as faker from 'faker';
import { MatchControlComponent } from './match-control.component';

describe('MatchControlComponent', () => {
  let component: MatchControlComponent;
  let fixture: ComponentFixture<MatchControlComponent>;
  let renderer: Renderer2;
  let elementRef: ElementRef;
  let checkboxElementRef: ElementRef;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [MatchControlComponent]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(MatchControlComponent);
    component = fixture.componentInstance;
    component.formControlName = 'match-control';
    renderer = fixture.componentRef.injector.get<Renderer2>(Renderer2 as Type<Renderer2>);
    elementRef = fixture.elementRef;
    checkboxElementRef = fixture.componentInstance.checkboxElementRef;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('uniqueFormControlId', () => {
    it('should return a id containing the formControlName and the given id', () => {
      component.formControlName = faker.random.word();
      component.id = faker.random.uuid();

      expect(component.uniqueFormControlId).toBe(`${component.formControlName}-${component.id}`);
    });
  });

  describe('registerOnChange', () => {
    it('should set the onChangeCallback', () => {
      const fakeCallback = jest.fn();
      component.registerOnChange(fakeCallback);

      expect(component.onChangeCallback).toBe(fakeCallback);
    });
  });

  describe('registerOnTouched', () => {
    it('should set the onTouchedCallback', () => {
      const fakeCallback = jest.fn();
      component.registerOnTouched(fakeCallback);

      expect(component.onTouchedCallback).toBe(fakeCallback);
    });
  });

  describe('writeValue', () => {
    it('should set the ckecked property of the elementRef', () => {
      const value = faker.random.boolean();
      renderer.setProperty = jest.fn();

      component.writeValue(value);

      expect((renderer.setProperty as jest.Mock).mock.calls[0]).toEqual([
        checkboxElementRef.nativeElement,
        'checked',
        value
      ]);

      expect((renderer.setProperty as jest.Mock).mock.calls[1]).toEqual([
        checkboxElementRef.nativeElement,
        'disabled',
        value
      ]);

      (renderer.setProperty as jest.Mock).mockClear();
    });
  });

  describe('onChange', () => {
    it('should call the onChangeCallback function', () => {
      const value = faker.random.boolean();
      component.onChangeCallback = jest.fn();

      component.onChange(value);

      expect(component.onChangeCallback).toHaveBeenCalledWith(value);
    });

    it('should call the onChange event handler if the checkbox changed', () => {
      const checked = faker.random.boolean();
      const checkboxElement = fixture.debugElement.query(By.css('[data-test-id="match-checkbox"]'));
      component.onChange = jest.fn();

      checkboxElement.triggerEventHandler('change', {
        target: {
          checked
        }
      });

      expect(component.onChange).toHaveBeenCalledWith(checked);
    });
  });

  describe('onTouched', () => {
    it('should call the onTouchedCallback function', () => {
      component.onTouchedCallback = jest.fn();

      component.onTouched();

      expect(component.onTouchedCallback).toHaveBeenCalledWith();
    });

    it('should call the onTouched event handler if the checkbox gets blured', () => {
      const checkboxElement = fixture.debugElement.query(By.css('[data-test-id="match-checkbox"]'));
      component.onTouched = jest.fn();

      checkboxElement.triggerEventHandler('blur', {});

      expect(component.onTouched).toHaveBeenCalled();
    });
  });
});
