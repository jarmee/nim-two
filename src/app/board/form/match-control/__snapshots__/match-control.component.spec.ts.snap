// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchControlComponent should match the snapshot 1`] = `
<app-match-control
  elementRef={[Function ElementRef]}
  formControlName={[Function String]}
  renderer={[Function DebugRenderer2]}
>
  <input
    data-test-id="match-checkbox"
    id="match-control"
    name="match-control"
    type="checkbox"
  />
</app-match-control>
`;
